components/ai_chat/browser/ai_chat_tab_helper.cc

```diff
@@ -0,0 +1,461 @@
+    AddToConversationHistory({CharacterType::ASSISTANT,
+                              ConversationTurnVisibility::VISIBLE,
+                              updated_text});
+  } else {
+    chat_history_.back().text = updated_text;
+  }
+
+  // Trigger an observer update to refresh the UI.
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+  }
+}
+
+void AIChatTabHelper::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+}
+
+void AIChatTabHelper::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void AIChatTabHelper::GeneratePageText() {
+  DCHECK(HasUserOptedIn());
+  auto* primary_rfh = web_contents()->GetPrimaryMainFrame();
+
+  if (!primary_rfh) {
+    VLOG(1) << "Summary request submitted for a WebContents without a "
+               "primary main frame";
+    return;
+  }
+  FetchPageContent(
+      web_contents(),
+      base::BindOnce(&AIChatTabHelper::OnTabContentRetrieved,
+                     weak_ptr_factory_.GetWeakPtr(), current_navigation_id_));
+}
+
+void AIChatTabHelper::OnTabContentRetrieved(int64_t for_navigation_id,
+                                            std::string contents_text,
+                                            bool is_video) {
+  if (for_navigation_id != current_navigation_id_) {
+    VLOG(1) << __func__ << " for a different navigation. Ignoring.";
+    return;
+  }
+  if (contents_text.empty()) {
+    VLOG(1) << __func__ << ": No data";
+    return;
+  }
+  is_video_ = is_video;
+  article_text_ = contents_text;
+
+  // Prevent indirect prompt injections being sent to the AI model.
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, ai_chat::kHumanPrompt,
+                                     "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, ai_chat::kAIPrompt, "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "<article>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "</article>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "<history>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "</history>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "<question>", "");
+  base::ReplaceSubstringsAfterOffset(&contents_text, 0, "</question>", "");
+
+  // Now that we have article text, we can suggest to summarize it
+  DCHECK(suggested_questions_.empty())
+      << "Expected suggested questions to be clear when there has been no"
+      << " previous text content but there were " << suggested_questions_.size()
+      << " suggested questions: "
+      << base::JoinString(suggested_questions_, ", ");
+
+  // Now that we have content, we can provide a summary on-demand. Add that to
+  // suggested questions.
+  // TODO(petemill): translation for this question
+  suggested_questions_.emplace_back(is_video_ ? "Summarize this video"
+                                              : "Summarize this page");
```

Comment:
Will address these strings in a follow-up. I think we'll do it completely different. Right now this whole feature assumes English only as the prompts are in English.